src/mpid/ch4/shm/xpmem/xpmem_avl.h

```diff
@@ -0,0 +1,564 @@
+
+    MPIDI_XPMEM_seg_avl_tree_init:
+        This function initializes the basic attributes of AVL tree
+        including root node, tree size, pool count, node pool, and it
+        should be called only once for a AVL tree at the beginning of
+        the program.
+
+    MPIDI_XPMEM_seg_avl_tree_search:
+        This function performs binary search to find matched segment.
+        Assume we have an input segment from user, this function will
+        find a matched segment if the input segment low address and high
+        address locate between the recorded segment in AVL tree. If low
+        address of input segment is smaller than recorded segment, we
+        need to search left branch; or else, we need to search right
+        branch.
+
+    MPIDI_XPMEM_seg_avl_tree_insert:
+        This function implements basic AVL tree insertion operation.
+        The key we use to decide the order of segment is the low address
+        of segment. If low address of a new segment is smaller than the
+        current segment node in AVL tree, it will be inserted into left
+        branch; or else, it will be inserted into right branch. After
+        normal insertion, we check the height of each sub tree to make
+        sure all of sub tree still obey the AVL tree requirement (left
+        branch and right branch height difference cannot exceed 1). If
+        it exceeds this limit, we need to adjust the height of tree to
+        make it balance and satisfy the requirement of AVL tree again.
+
+    MPIDI_XPMEM_seg_avl_tree_delete_all:
+        This function checks reference count of node and output warning
+        info if reference count is larger than 0, then it deletes and
+        releases all the node buffer pools allocated.
+
+*/
+
+
+/****************************************/
+/* AVL tree init related functions */
+/****************************************/
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_seg_avl_tree_init
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_seg_avl_tree_init(MPIDI_XPMEM_avl_tree_t * tree)
+{
+    int i;
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_SEG_AVL_TREE_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_SEG_AVL_TREE_INIT);
+
+    tree->root = NULL;
+    tree->tree_size = 0;
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_SEG_AVL_TREE_INIT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+
+/****************************************/
+/* AVL tree delete related functions */
+/****************************************/
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_avl_tree_node_ref_check
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_avl_tree_node_ref_check(MPIDI_XPMEM_avl_tree_t * tree)
+{
+    int ret;
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_seg_t *node_stack[64];
+    int addr_stack[64];
+    int sp = 0;
```

Comment:
Please use a macro to control the temporary stack size. The current code is hard to maintain.

Besides, what happens if `sp>=64` ? Can we just allocate the stack with the height of the tree as the size ? 

Same question for the `insert` routine.

Reply:
AVL tree is a balanced tree. It is impossible to have 64 height. 2^64 is a crazy large number, so 64 stack size is totally enough.
I will use a macro to operate stack.

Reply:
Sounds good, please also have a short comment explaining the above reason.